flink-runtime/src/test/java/org/apache/flink/runtime/failure/FailureEnricherUtilsTest.java

```diff
@@ -0,0 +1,284 @@
+        }
+    }
+
+    @Test
+    public void testLabelFailureWithInvalidEnricher() {
+        // validate labelFailure by enricher with wrong outputKeys
+        final Throwable cause = new RuntimeException("test exception");
+        final String invalidEnricherKey = "invalidKey";
+        final Set<FailureEnricher> failureEnrichers = new HashSet<>();
+        final FailureEnricher invalidEnricher =
+                new TestEnricher(
+                        Collections.singletonMap(invalidEnricherKey, "enricherValue"),
+                        "enricherKey");
+        failureEnrichers.add(invalidEnricher);
+
+        final CompletableFuture<Map<String, String>> result =
+                FailureEnricherUtils.labelFailure(cause, null, failureEnrichers);
+
+        try {
+            // Ignoring labels
+            final Map<String, String> labels = result.get();
+            assertThat(labels).isEmpty();
+        } catch (Exception e) {
+            fail("This should never happen.");
+        }
+    }
+
+    @Test
+    public void testLabelFailureMergeException() {
+        // Throwing exception labelFailure when merging duplicate keys
+        final Throwable cause = new RuntimeException("test failure");
+        final FailureEnricher firstEnricher = new TestEnricher("key1", "key2");
+        final FailureEnricher secondEnricher = new TestEnricher("key2", "key3");
+        final Set<FailureEnricher> enrichers =
+                new HashSet<FailureEnricher>() {
+                    {
+                        add(firstEnricher);
+                        add(secondEnricher);
+                    }
+                };
+
+        final CompletableFuture<Map<String, String>> result =
+                FailureEnricherUtils.labelFailure(cause, null, enrichers);
+
+        try {
+            result.get();
+        } catch (Exception e) {
+            assertThat(e).hasMessageContaining(String.format(MERGE_EXCEPTION_MSG, "key2"));
+        }
+    }
+
+    /**
+     * Testing plugin manager for {@link FailureEnricherFactory} utilizing {@link
+     * TestFailureEnricherFactory}.
+     *
+     * @return the testing PluginManager
+     */
+    private PluginManager getPluginManager() {
+        final Map<Class<?>, Iterator<?>> plugins = new HashMap<>();
+        plugins.put(
+                FailureEnricherFactory.class,
+                IteratorUtils.singletonIterator(new TestFailureEnricherFactory()));
+        return new TestingPluginManager(plugins);
+    }
+
+    /** Factory implementation of {@link TestEnricher} used for plugin load testing. */
+    private class TestFailureEnricherFactory implements FailureEnricherFactory {
+
+        @Override
+        public FailureEnricher createFailureEnricher(Configuration conf) {
+            return new TestEnricher();
+        }
+    }
+
+    private static class AndAnotherTestEnricher extends TestEnricher {
+
+        public AndAnotherTestEnricher(String... outputKeys) {
```

Comment:
nit: public constructors don't really make sense with private classes, we usually just use package private for conveniece

```suggestion
       AndAnotherTestEnricher(String... outputKeys) {
```